### PR TITLE
Update DlNNE engine 

### PR DESCRIPTION
**Pull request by denglin-github**
### PR types

New features
### PR changes

Others
### Describe

add more config param for AnalysisConfig.enable_dlnne()  && support calibration mode for dlnne run_time

### CODE REVIEW

File: paddle/fluid/inference/analysis/ir_passes/dlnne_subgraph_pass.cc [hunk 1]
```diff
@@ -11,87 +11,333 @@
...
 #include <set>
 
+#include "paddle/fluid/framework/attribute.h"
 #include "paddle/fluid/framework/ir/graph_pattern_detector.h"
 #include "paddle/fluid/framework/ir/subgraph_detector.h"
 #include "paddle/fluid/framework/op_version_registry.h"
+#include "paddle/fluid/framework/type_defs.h"
 #include "paddle/fluid/inference/analysis/helper.h"
-#include "paddle/fluid/inference/analysis/ir_passes/dlnne_reg_py.h"
+#include "paddle/fluid/inference/analysis/ir_passes/dlnne_subgraph_pass.h"
```
Comment (zhoutianzi666) [inline paddle/fluid/inference/analysis/ir_passes/dlnne_subgraph_pass.cc:26]: dlnne_reg_py.h这个文件还要吗？不要的话就删了吧。

Comment (denglin-github) [reply paddle/fluid/inference/analysis/ir_passes/dlnne_subgraph_pass.cc:26]: OK, Done, thanks~

File: paddle/fluid/inference/analysis/ir_passes/dlnne_subgraph_pass.cc [hunk 2]
```diff
@@ -153,6 +399,46 @@ std::string replace_name(std::string name,
...
+          copy(sp.begin(),
+               sp.end(),
+               std::ostream_iterator<int64_t>(sp_str, ","));
+
+          LOG(INFO)
+              << "warning: fix var:" << name << " batch,shape is ["
+              << sp_str.str()
+              << "],we assume subgraph all inputs/outputs first dim is batch,"
+                 "so sometime is error,but ok on some network,wo suggest you "
+                 "use fix shape's main graph...."
```
Comment (qingqing01) [inline paddle/fluid/inference/analysis/ir_passes/dlnne_subgraph_pass.cc]: 语法有误:  Warning，大写字母开头，后面提示信息有很多语法问题~

File: paddle/fluid/inference/analysis/ir_passes/dlnne_subgraph_pass.cc [hunk 3]
```diff
@@ -153,6 +399,46 @@ std::string replace_name(std::string name,
...
+              << std::endl;
```
Comment (qingqing01) [inline paddle/fluid/inference/analysis/ir_passes/dlnne_subgraph_pass.cc]: 这行也不需要吧

File: paddle/fluid/inference/analysis/ir_passes/dlnne_subgraph_pass.cc [hunk 4]
```diff
@@ -11,87 +11,333 @@
...
   }
 
-  return 0;
-}
-int ConvertGraph(std::string graph_name) {
-  LOG(INFO) << "starting doing convert_graph";
+  std::vector<framework::OpDesc *> all_ops = block->AllOps();
+
+  for (size_t i = 0; i < block->OpSize(); i++) {
+    // infer shape for bilinear_interp_v2
```
Comment (qingqing01) [inline paddle/fluid/inference/analysis/ir_passes/dlnne_subgraph_pass.cc]: 增加注释这里为啥要特殊处理  bilinear_interp_v2

File: paddle/fluid/inference/analysis/ir_passes/dlnne_subgraph_pass.cc [hunk 5]
```diff
@@ -11,87 +11,333 @@
...
+      "transpose2",
+      "fill_constant_batch_size_like",
+      "flatten_contiguous_range",
+      "batch_norm",
+      "unsqueeze2",
+  };
+
+  static std::unordered_set<std::string> dynamic_pass_op_set{
+      "reshape2",
+  };
```
Comment (qingqing01) [inline paddle/fluid/inference/analysis/ir_passes/dlnne_subgraph_pass.cc:295]: 增加注释，这里为啥把Op分了多类

File: paddle/fluid/operators/dlnne/dlnne_engine_op.cc [hunk 6]
```diff
@@ -41,7 +143,15 @@ class DlnneEngineOpMaker : public framework::OpProtoAndCheckerMaker {
     AddAttr<std::string>(
         "engine_key",
         "The engine_key here is used to distinguish different DLNNE Engines");
-    AddAttr<framework::BlockDesc*>("sub_block", "the trt block");
+    AddAttr<int32_t>("max_batch_size", "engine max_batch_size");
+    AddAttr<bool>("use_static_batch", "static batch fix for [?,H,W,C]");
+    AddAttr<std::string>("weight_share_mode", "dlnne weight_share_mode");
```
Comment (qingqing01) [inline paddle/fluid/operators/dlnne/dlnne_engine_op.cc]: weight_share_mode都有那几种，注释需要写的明白一些

File: paddle/fluid/operators/dlnne/dlnne_engine_op.cc [hunk 7]
```diff
@@ -41,7 +143,15 @@ class DlnneEngineOpMaker : public framework::OpProtoAndCheckerMaker {
...
+    AddAttr<bool>("use_calib_mode", "dlnne use calib mode");
+    AddAttr<bool>("enable_int8", "dlnne enable int8");
+    AddAttr<bool>("calibration_mode", "dlnne calibration_mode");
```
Comment (qingqing01) [inline paddle/fluid/operators/dlnne/dlnne_engine_op.cc:156]: use_calib_mode和calibration_mode两个都是bool，需要2个变量吗？

File: paddle/fluid/operators/dlnne/dlnne_engine_op.h [hunk 8]
```diff
@@ -97,81 +186,213 @@ class DlnneEngineOp : public framework::OperatorBase {
...
 
-    for (std::string &output_name : output_names) {
-      int BindIndex = engine->GetBindingIndex(output_name.c_str());
-      InputIndexToBindIndex_.push_back(BindIndex);
-    }
+      inference::ConvertPaddle2Onnx(subgraph_root_path_, engine_key_);
+      std::string onnx_file_name =
+          subgraph_root_path_ + "/" + engine_key_ + ".onnx";
+      std::string rlym_file_name =
+          subgraph_root_path_ + "/" + engine_key_ + ".rlym";
```
Comment (qingqing01) [inline paddle/fluid/operators/dlnne/dlnne_engine_op.h:279]: onnx-file的名字还是传入ConvertPaddle2Onnx好些吧，不用多个地方都去创建

File: paddle/fluid/operators/dlnne/dlnne_engine_op.cc [hunk 9]
```diff
@@ -28,6 +28,108 @@ void CopyTensorCpuToDevice(void* dst_ptr, void* src_ptr, int total_bytes) {
...
+
+  std::shuffle(str.begin(), str.end(), generator);
+  return str.substr(0, 32);
+}
+
+void ConvertPaddle2Onnx(std::string subgraph_root_path,
+                        std::string engine_key) {
+  std::string onnx_file_name = subgraph_root_path + "/" + engine_key + ".onnx";
+  // std::fstream onnx_file;
+  // onnx_file.open(onnx_file_name.c_str(), std::ios::in);
```
Comment (qingqing01) [inline paddle/fluid/operators/dlnne/dlnne_engine_op.cc]: 删除没用的代码

File: paddle/fluid/operators/dlnne/dlnne_engine_op.cc [hunk 10]
```diff
@@ -28,6 +28,108 @@ void CopyTensorCpuToDevice(void* dst_ptr, void* src_ptr, int total_bytes) {
...
+  if (!FileExists(onnx_file_name.c_str())) {
+    std::stringstream convert_cmd;
+    convert_cmd << "paddle2onnx --model_dir " << subgraph_root_path
+                << " --save_file " << onnx_file_name << " --opset_version 11";
```
Comment (qingqing01) [inline paddle/fluid/operators/dlnne/dlnne_engine_op.cc:90]: 这里用户还得安装paddle2onnx才能成功

File: paddle/fluid/operators/dlnne/dlnne_engine_op.h [hunk 11]
```diff
@@ -318,36 +612,142 @@ class DlnneEngineOp : public framework::OperatorBase {
...
         bind_index++;
       }
       cudaFree(engine_input_ptr[InputIndexToBindIndex_[i]]);
     }
+    dlnne_create_lock.unlock();
+  }
+
+  void RunNativeImpl(const framework::Scope &scope,
+                     const platform::Place &dev_place) const {
+    LOG(INFO) << "RunNativeImpl";
```
Comment (qingqing01) [inline paddle/fluid/operators/dlnne/dlnne_engine_op.h]: 一些无意义的log可以去掉，或用VLOG

File: paddle/fluid/operators/dlnne/dlnne_engine_op.h [hunk 12]
```diff
@@ -318,36 +612,142 @@ class DlnneEngineOp : public framework::OperatorBase {
...
+      // engine, there is same code
+      auto t_shape = phi::vectorize<int64_t>(t.dims());
+      std::vector<int64_t> input_shape(t_shape.begin(), t_shape.end());
+      calib_data_shape_map.emplace(x, input_shape);
+      std::string data_type = inference::ConvertType(t.type());
+      calib_data_type_map.emplace(x, data_type);
+
+      int data_bytes = inference::GetDataByte(t.type());
+      LOG(INFO) << "input name: " << x << ", data_type: " << data_type;
+      LOG(INFO) << "data shape: ";
```
Comment (qingqing01) [inline paddle/fluid/operators/dlnne/dlnne_engine_op.h]: 同上~

**Review by qingqing01 — COMMENTED** [review]
最好增加单测

File: paddle/fluid/inference/analysis/ir_passes/dlnne_subgraph_pass.cc (hunk 2 above)
Comment (denglin-github) [reply paddle/fluid/inference/analysis/ir_passes/dlnne_subgraph_pass.cc]: Done, thanks

File: paddle/fluid/inference/analysis/ir_passes/dlnne_subgraph_pass.cc (hunk 3 above)
Comment (denglin-github) [reply paddle/fluid/inference/analysis/ir_passes/dlnne_subgraph_pass.cc]: Done, thanks

File: paddle/fluid/inference/analysis/ir_passes/dlnne_subgraph_pass.cc (hunk 4 above)
Comment (denglin-github) [reply paddle/fluid/inference/analysis/ir_passes/dlnne_subgraph_pass.cc]: annotation has  added, thanks

File: paddle/fluid/inference/analysis/ir_passes/dlnne_subgraph_pass.cc (hunk 5 above)
Comment (denglin-github) [reply paddle/fluid/inference/analysis/ir_passes/dlnne_subgraph_pass.cc:295]: have added explain annotation,   thanks

File: paddle/fluid/operators/dlnne/dlnne_engine_op.cc (hunk 6 above)
Comment (denglin-github) [reply paddle/fluid/operators/dlnne/dlnne_engine_op.cc]: Done, thanks

File: paddle/fluid/operators/dlnne/dlnne_engine_op.cc (hunk 7 above)
Comment (denglin-github) [reply paddle/fluid/operators/dlnne/dlnne_engine_op.cc:156]: both of them are useful, add annotation of them

File: paddle/fluid/operators/dlnne/dlnne_engine_op.h (hunk 8 above)
Comment (denglin-github) [reply paddle/fluid/operators/dlnne/dlnne_engine_op.h:279]: Done, thanks

File: paddle/fluid/operators/dlnne/dlnne_engine_op.cc (hunk 9 above)
Comment (denglin-github) [reply paddle/fluid/operators/dlnne/dlnne_engine_op.cc]: Done, thanks

File: paddle/fluid/operators/dlnne/dlnne_engine_op.cc (hunk 10 above)
Comment (denglin-github) [reply paddle/fluid/operators/dlnne/dlnne_engine_op.cc:90]: the custormer use it in a python virtual_env, which pip has installed paddlepaddle and paddle2onnx

File: paddle/fluid/operators/dlnne/dlnne_engine_op.h (hunk 11 above)
Comment (denglin-github) [reply paddle/fluid/operators/dlnne/dlnne_engine_op.h]: Done, thanks

File: paddle/fluid/operators/dlnne/dlnne_engine_op.h (hunk 12 above)
Comment (denglin-github) [reply paddle/fluid/operators/dlnne/dlnne_engine_op.h]: Done, thanks

**Review by qingqing01 — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Review by leiqing1 — APPROVED** [review]

**Pull request merged**